Intermediate chapter on Python scripting (#232)
merely-useful/merely-useful.github.io: rse-py-scripting.Rmd

# Command line programs {#rse-py-scripting}

## Questions {#rse-py-scripting-questions}

```{r, child="questions/rse-py-scripting.md"}
```

## Introduction {#rse-py-scripting-intro}

The Jupyter Notebook, PyCharm and other interactive tools
are great for prototyping code and exploring data,
but in many cases you will ultimately want to take the code you've developed
using those tools and apply it to thousands of data files, 
run it with many different parameters,
and/or combine it with other programs in a data analysis pipeline. 

> In order to do these things,

Reviewer: ```suggestion
To effectively perform these tasks,
```

> it is often necessary/preferable to turn the code you've developed into a program 

Reviewer: ```suggestion
it is often necessary to turn the code you've developed into a program 
```

> that works like any other Unix command-line tool.

Reviewer: ```suggestion
that can be run from the Unix shell as a standalone program just like other command-line tools.
```

In this chapter we take the code we developed earlier (TODO: Link)
and turn it into a Python script (i.e. a command line program written in Python).

## A template script {#rse-py-scripting-template}

In general, the first thing that gets added to any Python script is the following:

```python
if __name__ == '__main__':
    main()
```

The reason we need these two lines of code

> is that running a Python script in bash is very similar to importing that file in Python. 

Reviewer: I remember struggling with this when I started learning Python... I think we could make these two paragraphs slightly clearer, what do you think of something along these lines?

```suggestion
The reason we need these two lines of code
is to differentiate between running a Python file as a standalone program
and importing it as a module. 
When you import a Python file as a module,
the `__name__` variable is automatically set to the name of the file
(e.g. when importing script.py, `__name__` is `script`).
When running a Python file as a standalone program,
`__name__` is always set to `__main__`.
This means that we can separate the two cases above by checking the value of `__name__`.
When the Python file is running as a standalone program,
you want to execute its main functionality.
You conventionally define what this is inside a function named `main()`,
but you could call this function whatever you prefer.
```

The biggest difference is that we don’t expect anything to happen when we import a file, 
whereas when running a script we expect to see some output
(e.g. an output file, figure and/or some text printed to the screen).

The `__name__` variable exists to handle these two situations.
When you import a Python file `__name__` is set to the name of that file
(e.g. when importing script.py, `__name__` is `script`),
but when running a script in bash `__name__` is always set to `__main__`.
The convention is to call the function that produces the output `main()`,
but you can call it whatever you like.

The next thing you'll need is a library to parse the command line for input arguments.
The most widely used option is 
[argparse](https://docs.python.org/3/library/argparse.html).

Putting those together,
here's a basic python script for us to start with:

```shell

> $ cat script_template.py

Reviewer: Do we need this code block? Can we say "Create the file `script_template.py`" instead?

if __name__ == '__main__':

    description='Print the input arguments to the screen.'
    parser = argparse.ArgumentParser(description=description)
    
    parser.add_argument("infile", type=str, help="Input file name")
    parser.add_argument("outfile", type=str, help="Output file name")

    args = parser.parse_args()            
    main(args)
```

By running `script_template.py` at the command line
we'll see that `argparse` handles all the input arguments:

```shell
$ python script_template.py in.csv out.png
```

```text
Input file:  in.csv
Output file:  out.png
```

It also generates help information for the user:

```shell
$ python script_template.py -h
```

```text
usage: script_template.py [-h] infile outfile

Print the input arguments to the screen.

positional arguments:
  infile      Input file name
  outfile     Output file name

optional arguments:
  -h, --help  show this help message and exit
```

and issues errors when users give the program invalid arguments:

```shell
$ python script_template.py in.csv
```

```text
usage: script_template.py [-h] infile outfile
script_template.py: error: the following arguments are required: outfile
```

> ## Adding our code {#rse-py-scripting-code}

Reviewer: Should we cover optional and default arguments before this or it that outside the scope of this book?

Author: I think in the process of "adding our code" we should introduce optional and default arguments.
